AceState-3.0/AceState-3.0.lua

```diff
@@ -0,0 +1,110 @@
+local MAJOR, MINOR = "AceState-3.0", 1
+local AceState, oldminor = LibStub:NewLibrary(MAJOR, MINOR)
+
+if not AceState then
+	return -- No upgrade needed
+end
+
+-- Register a reducer with the initial state and the reducer function
+function AceState:RegisterReducer(name, initialState, reducer)
+	-- If the reducer is a string, it should be a method on the addon
+	if type(reducer) == "string" then
+		if not self[reducer] or type(self[reducer]) ~= "function" then
+			error("Reducer method " .. reducer .. " does not exist on addon")
+		end
+		reducer = self[reducer]
+	elseif type(reducer) ~= "function" then
+		error("Reducer must be a function or a string referring to a method on the addon.")
+	end
+
+	-- Return a dispatch function for this reducer
+	local dispatch = function(actionType, payload)
+		local oldState = self.DeepCopy(self.state[name])
+		local newState = self.reducers[name](oldState, actionType, payload)
+
+		-- Update the state
+		for k, v in pairs(newState) do
+			self.state[name][k] = v
+		end
+		for k in pairs(self.state[name]) do
+			if newState[k] == nil then
+				self.state[name][k] = nil
+			end
+		end
```

Comment:
This bit here is logic that transplants the updated state properties onto the original state object. By transplanting properties onto the original object we ensure that any references to that original object remain intact. If we simply overwrote the original object with the new state then any existing references to the original object would be invalid.